small/obuf_asan.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "obuf.h"
+#include "util.h"
+
+#include <string.h>
+
+void
+obuf_create(struct obuf *buf, struct slab_cache *slabc, size_t start_capacity)
+{
+	buf->slabc = slabc;
+	buf->start_capacity = start_capacity;
+	size_t size = (SMALL_OBUF_IOV_MAX + 1) * sizeof(*buf->iov);
+	buf->iov = small_xmalloc(size);
+	memset(buf->iov, 0, size);
+	memset(buf->capacity, 0,
+	       lengthof(buf->capacity) * sizeof(buf->capacity[0]));
+	buf->pos = 0;
+	buf->used = 0;
+	buf->reserved = 0;
+}
+
+/** Allocate new memory block whether for allocation or reservation. */
+static void *
+obuf_prepare_buf(struct obuf *buf, size_t size)
+{
+	if (buf->pos >= SMALL_OBUF_IOV_CHECKED_SIZE - 1) {
+		int gpos = buf->pos - SMALL_OBUF_IOV_CHECKED_SIZE;
+		if (gpos < 0 ||
+		    (buf->iov[buf->pos].iov_len + size) > buf->capacity[gpos]) {
```

Comment:
`gpos < 0` looks confusing. Intuitively, the code should look like this:

```C
if (buf->pos >= SMALL_OBUF_IOV_CHECKED_SIZE) {
       if (buf->iov[buf->pos].iov_len + size > buf.capacity[gpos]) {
           // allocate a new block
       }
       // allocate from the last block
}
```

Also, `gpos` is a bad name - it doesn't give any clue about what this variable denotes.

Overall, the code is difficult for understanding because the "checked" vectors. Maybe, better maintain the capacity array for both checked and unchecked allocations to keep things simple? Then there wouldn't be this complex conditional here and you wouldn't have to check the `pos` against `SMALL_OBUF_IOV_CHECKED_SIZE` in `obuf_reserve`.

Reply:
Done. Additionally

- made further code simplifications
- enhance poison (for geometric part) 

<details>
<summary>diff</summary>

```diff
commit 2a8e552ceb9313c1bdfa41dd5ca40a19f86cb648
Author: Nikolay Shirokovskiy <nshyrokovskiy@gmail.com>
Date:   Mon Sep 25 17:11:58 2023 +0300

    TMP: simplify code

diff --git a/include/small/obuf_asan.h b/include/small/obuf_asan.h
index a84efe7..a05bbdb 100644
--- a/include/small/obuf_asan.h
+++ b/include/small/obuf_asan.h
@@ -44,18 +44,12 @@ enum {
 	 * NULL terminated thus maximum index holding vector is one less.
 	 */
 	SMALL_OBUF_IOV_MAX = 128,
-	/**
-	 * Number of exponentially growing vectors at the end of iovec
-	 * vectors list.
-	 */
-	SMALL_OBUF_IOV_GEOMETRIC_SIZE = 32,
 	/**
 	 * Number of vectors at the beginning of iovec vector list
 	 * that are allocated with distinct malloc call. Let's refer
 	 * them as checked vectors.
 	 */
-	SMALL_OBUF_IOV_CHECKED_SIZE = SMALL_OBUF_IOV_MAX + 1 -
-				      SMALL_OBUF_IOV_GEOMETRIC_SIZE,
+	SMALL_OBUF_IOV_CHECKED_SIZE = 96,
 	/** Alignment for the checked vectors. */
 	SMALL_OBUF_ALIGNMENT = 1,
 	/** Minimum size when reserving memory. See struct obuf comment. */
@@ -101,11 +95,8 @@ struct obuf {
 	 * is requested.
 	 */
 	size_t start_capacity;
-	/**
-	 * Capacity of exponentially growing vectors. Note that size is off
-	 * by one because iov list is NULL terminated.
-	 */
-	size_t capacity[SMALL_OBUF_IOV_GEOMETRIC_SIZE - 1];
+	/** How many bytes are actually allocated for each iovec. */
+	size_t capacity[SMALL_OBUF_IOV_MAX];
 	/**
 	 * If pos == 0 and iov[0] == NULL then obuf is empty. Otherwise
 	 * pos is index of the vector in the vectors list holding last
@@ -115,8 +106,8 @@ struct obuf {
 	int pos;
 	/** Total size of allocations. */
 	size_t used;
-	/** If not 0 then amount of memory reserved with the obuf_reserve. */
-	size_t reserved;
+	/** Whether memory is reserved with the obuf_reserve. */
+	bool reserved;
 };
 
 void
diff --git a/small/obuf_asan.c b/small/obuf_asan.c
index 26a8c0d..94a5131 100644
--- a/small/obuf_asan.c
+++ b/small/obuf_asan.c
@@ -43,84 +43,92 @@ obuf_create(struct obuf *buf, struct slab_cache *slabc, size_t start_capacity)
 	       lengthof(buf->capacity) * sizeof(buf->capacity[0]));
 	buf->pos = 0;
 	buf->used = 0;
-	buf->reserved = 0;
+	buf->reserved = false;
 }
 
-/** Allocate new memory block whether for allocation or reservation. */
-static void *
+/**
+ * Pointer to unallocated space. This is free space in the last allocated
+ * buffer in buf->iov that is not allocated by user.
+ */
+static char *
+obuf_unallocated_ptr(struct obuf *buf)
+{
+	return buf->iov[buf->pos].iov_base + buf->iov[buf->pos].iov_len;
+}
+
+/** Size of unallocated space. See obuf_unallocated_ptr. */
+static size_t
+obuf_unallocated_size(const struct obuf *buf)
+{
+	return buf->capacity[buf->pos] - buf->iov[buf->pos].iov_len;
+}
+
+/** Make sure size bytes are available as continuous chunk. */
+static void
 obuf_prepare_buf(struct obuf *buf, size_t size)
 {
-	if (buf->pos >= SMALL_OBUF_IOV_CHECKED_SIZE - 1) {
-		int gpos = buf->pos - SMALL_OBUF_IOV_CHECKED_SIZE;
-		if (gpos < 0 ||
-		    (buf->iov[buf->pos].iov_len + size) > buf->capacity[gpos]) {
-			size_t capacity = buf->start_capacity << (gpos + 1);
-			while (capacity < size)
-				capacity <<= 1;
+	if (buf->pos < SMALL_OBUF_IOV_CHECKED_SIZE) {
+		/* See limit explanation in the header. */
+		if (size < SMALL_OBUF_MIN_RESERVE)
+			size = SMALL_OBUF_MIN_RESERVE;
+
+		char *alloc = small_asan_alloc(size, SMALL_OBUF_ALIGNMENT, 0);
+		char *payload = small_asan_payload_from_header(alloc);
+
+		/* See obuf.pos semantics in struct definition. */
+		if (buf->iov[buf->pos].iov_base != NULL)
 			buf->pos++;
-			small_asan_assert(buf->pos < SMALL_OBUF_IOV_MAX);
-			struct iovec *iov = &buf->iov[buf->pos];
-			iov->iov_base = small_xmalloc(capacity);
-			buf->capacity[gpos + 1] = capacity;
-			iov->iov_len = 0;
-		}
-		struct iovec *iov = &buf->iov[buf->pos];
-		void *ptr = iov->iov_base + iov->iov_len;
-		return ptr;
+		buf->iov[buf->pos].iov_base = payload;
+		buf->iov[buf->pos].iov_len = 0;
+		buf->capacity[buf->pos] = size;
+		return;
 	}
 
-	char *alloc = small_asan_alloc(size, SMALL_OBUF_ALIGNMENT, 0);
-	char *payload = small_asan_payload_from_header(alloc);
+	size_t used = buf->iov[buf->pos].iov_len + size;
+	if (used > buf->capacity[buf->pos]) {
+		size_t capacity = buf->start_capacity;
+		capacity <<= buf->pos - SMALL_OBUF_IOV_CHECKED_SIZE;
+		while (capacity < size)
+			capacity <<= 1;
 
-	/* See obuf.pos semantics in struct definition. */
-	if (buf->iov[buf->pos].iov_base != NULL)
 		buf->pos++;
-	buf->iov[buf->pos].iov_base = payload;
-	buf->iov[buf->pos].iov_len = 0;
-
-	return payload;
+		small_asan_assert(buf->pos < SMALL_OBUF_IOV_MAX);
+
+		char *alloc = small_xmalloc(capacity);
+		buf->iov[buf->pos].iov_base = alloc;
+		buf->iov[buf->pos].iov_len = 0;
+		buf->capacity[buf->pos] = capacity;
+	} else {
+		ASAN_UNPOISON_MEMORY_REGION(obuf_unallocated_ptr(buf),
+					    obuf_unallocated_size(buf));
+	}
 }
 
 void *
 obuf_reserve(struct obuf *buf, size_t size)
 {
-	small_asan_assert(buf->reserved == 0);
-
-	/* See limit explanation in the header. */
-	if (buf->pos < SMALL_OBUF_IOV_CHECKED_SIZE - 1 &&
-	    size < SMALL_OBUF_MIN_RESERVE)
-		size = SMALL_OBUF_MIN_RESERVE;
+	small_asan_assert(!buf->reserved);
 
-	void *ptr = obuf_prepare_buf(buf, size);
-	buf->reserved = size;
-	return ptr;
-}
-
-/** Allocate memory in case of prior reservation. */
-void *
-obuf_alloc_reserved(struct obuf *buf, size_t size)
-{
-	small_asan_assert(size <= buf->reserved);
-
-	struct iovec *iov = &buf->iov[buf->pos];
-	char *ptr = iov->iov_base + iov->iov_len;
-	iov->iov_len += size;
-	buf->used += size;
-	if (buf->pos < SMALL_OBUF_IOV_CHECKED_SIZE)
-		ASAN_POISON_MEMORY_REGION(ptr + size, buf->reserved - size);
-	buf->reserved = 0;
-	return ptr;
+	obuf_prepare_buf(buf, size);
+	buf->reserved = true;
+	return obuf_unallocated_ptr(buf);
 }
 
 void *
 obuf_alloc(struct obuf *buf, size_t size)
 {
-	if (buf->reserved != 0)
-		return obuf_alloc_reserved(buf, size);
-
-	void *ptr = obuf_prepare_buf(buf, size);
+	if (buf->reserved) {
+		small_asan_assert(size <= obuf_unallocated_size(buf));
+		buf->reserved = false;
+	} else {
+		obuf_prepare_buf(buf, size);
+	}
+	void *ptr = obuf_unallocated_ptr(buf);
 	buf->iov[buf->pos].iov_len += size;
 	buf->used += size;
+
+	ASAN_POISON_MEMORY_REGION(obuf_unallocated_ptr(buf),
+				  obuf_unallocated_size(buf));
 	return ptr;
 }
 
@@ -139,27 +147,23 @@ obuf_rollback_to_svp(struct obuf *buf, struct obuf_svp *svp)
 	      svp->iov_len == 0 &&
 	      buf->iov[0].iov_base != NULL))
 		pos++;
-	int endpos = buf->pos;
-	if (endpos > SMALL_OBUF_IOV_CHECKED_SIZE - 1)
-		endpos = SMALL_OBUF_IOV_CHECKED_SIZE - 1;
-	for (int i = pos; i <= endpos; i++) {
-		void *alloc = small_asan_header_from_payload(
-							buf->iov[i].iov_base);
-		small_asan_free(alloc);
 
+	for (int i = pos; i <= buf->pos; i++) {
+		if (i <= SMALL_OBUF_IOV_CHECKED_SIZE) {
+			void *ptr = buf->iov[i].iov_base;
+			small_asan_free(small_asan_header_from_payload(ptr));
+		} else {
+			free(buf->iov[i].iov_base);
+		}
 	}
-	int startpos = pos;
-	if (startpos < SMALL_OBUF_IOV_CHECKED_SIZE)
-		startpos = SMALL_OBUF_IOV_CHECKED_SIZE;
-	for (int i = startpos; i <= buf->pos; i++) {
-		free(buf->iov[i].iov_base);
-		buf->capacity[i - SMALL_OBUF_IOV_CHECKED_SIZE] = 0;
-	}
-	memset(&buf->iov[pos], 0, sizeof(*buf->iov) * (buf->pos - pos + 1));
+
+	size_t erase_size = buf->pos - pos + 1;
+	memset(&buf->iov[pos], 0, sizeof(buf->iov[0]) * erase_size);
+	memset(&buf->capacity[pos], 0, sizeof(buf->capacity[0]) * erase_size);
 	buf->pos = svp->pos;
 	buf->used = svp->used;
 	buf->iov[buf->pos].iov_len = svp->iov_len;
-	buf->reserved = 0;
+	buf->reserved = false;
 }
 
 void
diff --git a/test/obuf.c b/test/obuf.c
index 66d62a5..4058901 100644
--- a/test/obuf.c
+++ b/test/obuf.c
@@ -138,16 +138,19 @@ obuf_poison(struct slab_cache *slabc)
 	struct obuf buf;
 	obuf_create(&buf, slabc, 16384);
 	size_t size_max = 2 * small_getpagesize();
-	for (int i = 0; i < 87; i++) {
+	for (int i = 0; i < 7777; i++) {
 		size_t size_r = 1 + rand() % size_max;
 		size_t size_a = 1 + rand() % size_r;
 		char *ptr_r = obuf_reserve(&buf, size_r);
 		fail_unless(ptr_r != NULL);
-		memset(ptr_r, 0, size_a);
+		size_t reserved = buf.capacity[buf.pos] -
+				  buf.iov[buf.pos].iov_len;
+		ptr_r[0] = 0;
+		ptr_r[reserved - 1] = 0;
+		fail_unless(__asan_address_is_poisoned(ptr_r + reserved));
 		char *ptr_a = obuf_alloc(&buf, size_a);
 		fail_unless(ptr_r == ptr_a);
-		for (char *p = ptr_r + size_a; p < ptr_r + size_r; p++)
-			fail_unless(__asan_address_is_poisoned(p));
+		fail_unless(__asan_address_is_poisoned(ptr_a + size_a));
 	}
 	obuf_destroy(&buf);
 	ok(true);
```

</details>